File: packages/agents/src/openai-agents.ts
/**
 * `kriya-agents/openai-agents` — govern an OpenAI Agents SDK (JS/TS) function tool with kriya.
 *
 * The OpenAI Agents SDK defines a function tool with
 * `tool({ name, description, parameters, execute })`, where the SDK parses the model's arguments
 * against `parameters` (a zod schema) and calls `execute(input, context?, details?)` with the parsed
 * object (verified 2026-07-22 against the installed `@openai/agents` — `FunctionTool.invoke` parses
 * the arg JSON, then calls `execute`). Governing it is governing that `execute` — this adapter is a
 * thin shim over the framework-agnostic {@link govern} with no dependency on `@openai/agents`: you
 * wrap your `execute`, then pass it to the SDK's `tool()` yourself, so name/parameters (what the
 * model sees) are unchanged.
 *
 * ```ts
 * import { tool } from "@openai/agents";
 * import { z } from "zod";
 * import { GovernClient } from "kriya-agents";
 * import { governExecute } from "kriya-agents/openai-agents";
 *
 * const client = new GovernClient({ actor: "openai-agents", auditLog: "…/.kriya/audit/oai.jsonl" });
 * const add = tool({
 *   name: "add", description: "Add two numbers",
 *   parameters: z.object({ a: z.number(), b: z.number() }),
 *   execute: governExecute(client, "add", async ({ a, b }) => `${a + b}`),
 * });
 * ```
 *
 * A denied / approval-refused / over-budget call throws {@link GovernDenied}; the SDK surfaces it as
 * that tool's error (or routes it through your tool's `errorFunction`), so the model can adapt.
 *
 * Honest ceiling (state it in your app too): in-process governance is cooperative — a hostile agent
 * process can skip it (that is what launch-under containment / B14 is for) — and this lane governs the
 * action tier + signs the receipt; it does not see the tool's own outbound network calls.
 */

import { type GovernClient, type GovernHooks, govern } from "./govern-client.js";

/** The OpenAI Agents SDK `execute` shape: `(input, context?, details?) => result`. */
export type OpenAIToolExecute<A, R> = (input: A, context?: unknown, details?: unknown) => R | Promise<R>;

/**
 * Wrap an OpenAI Agents SDK tool `execute` so every invocation is policy / approval / budget gated
 * and (when it runs) signs a receipt. `actionId` is the id policy rules match on — use the tool's
 * `name`. The `context`/`details` the SDK passes are forwarded to your original `execute` untouched.
 */
export function governExecute<A extends Record<string, unknown>, R>(
  client: GovernClient,
  actionId: string,
  execute: OpenAIToolExecute<A, R>,
  hooks?: GovernHooks,
): OpenAIToolExecute<A, R> {
  return (input: A, context?: unknown, details?: unknown): Promise<R> =>
    // Reuse the framework-agnostic core; bind the SDK's context/details per call so the original
    // execute still receives them. A non-allow decision throws GovernDenied (the SDK surfaces it).
    govern<A, R>(client, actionId, (params: A) => execute(params, context, details), hooks)(input);
}

File: bindings/python/src/kriya/agents/openai_agents.py
"""kriya.agents.openai_agents — govern an OpenAI Agents SDK (Python) function tool with kriya.

The OpenAI Agents SDK (Python) turns a function into a tool with `@function_tool`, producing a
`FunctionTool` whose `on_invoke_tool(ctx, input: str)` is the SDK's invocation seam — it parses the
model's JSON arguments and calls your function (verified 2026-07-22 against `openai-agents` 0.18.3).
Governing the tool is governing that seam: this adapter wraps `on_invoke_tool` so every call is
policy -> approval -> budget gated and (when it runs) signs a receipt, with the tool's name and JSON
schema untouched (the model sees an identical tool). No dependency on `agents` -- you build the tool,
then hand it to `govern_function_tool`.

```python
from agents import function_tool
from kriya.agents import GovernClient
from kriya.agents.openai_agents import govern_function_tool

client = GovernClient(actor="openai-agents", audit_log="~/.kriya/audit/oai.jsonl")

@function_tool
def add(a: int, b: int) -> str:
    "Add two numbers."
    return str(a + b)

govern_function_tool(client, add)   # same tool object, now governed
```

A denied / approval-refused / over-budget call raises :class:`kriya.agents.GovernDenied`; the SDK
surfaces it to the model as the tool's error, so it can adapt. Honest ceiling: in-process governance
is cooperative (a hostile process can skip it -- that is what launch-under containment / B14 is for);
it governs the action tier + signs, and does not see the tool's own egress.
"""

from __future__ import annotations

import json
from typing import Any, Optional

from . import GovernClient, GovernDenied

__all__ = ["govern_function_tool"]


def govern_function_tool(client: GovernClient, tool: Any, action_id: Optional[str] = None) -> Any:
    """Govern a built OpenAI Agents SDK `FunctionTool` in place (its `on_invoke_tool` is wrapped).

    Returns the same `tool` object so it can be used inline. `action_id` defaults to `tool.name` --
    the id your policy rules match on. The wrapped seam takes the SDK's raw JSON argument string; it
    is parsed for the receipt's params (the SDK re-parses it for the underlying function, unchanged).
    """
    resolved_id = action_id or getattr(tool, "name", None)
    if not resolved_id:
        raise ValueError("govern_function_tool: tool has no name and no action_id was given")
    original = tool.on_invoke_tool

    async def governed(ctx: Any, input_str: str) -> Any:
        params: Any
        try:
            params = json.loads(input_str) if input_str and input_str.strip() else {}
        except Exception:
            params = {}
        if not isinstance(params, dict):
            params = {"input": params}

        # NOTE: GovernClient is synchronous (fast local IPC to kriya-govern); the brief block inside
        # this coroutine is acceptable for the check/record round-trips.
        gate = client.check(resolved_id, params)
        if gate.decision != "allow":
            raise GovernDenied(resolved_id, gate.decision, gate.reason)
        success = False
        try:
            result = await original(ctx, input_str)
            success = True
            return result
        finally:
            client.record(resolved_id, params, success)

    tool.on_invoke_tool = governed
    return tool
